File: Aeds2/TP02/Disney/DisneyPilha.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX_LINE 1000
#define MAX_FIELD 100
#define MAX_LIST 20

typedef struct {
    char show_id[MAX_FIELD];
    char type[MAX_FIELD];
    char title[MAX_FIELD];
    char director[MAX_FIELD];
    char cast[MAX_LIST][MAX_FIELD];
    int cast_size;
    char country[MAX_FIELD];
    char date_added[MAX_FIELD];
    int release_year;
    char rating[MAX_FIELD];
    char duration[MAX_FIELD];
    char listed_in[MAX_LIST][MAX_FIELD];
    int listed_in_size;
} Show;

typedef struct No {
    Show elemento;
    struct No* proximo;
} No;

// Função para inicializar um show vazio
Show criar_show_vazio() {
    Show s;
    strcpy(s.show_id, "NaN");
    strcpy(s.type, "NaN");
    strcpy(s.title, "NaN");
    strcpy(s.director, "NaN");
    s.cast_size = 0;
    strcpy(s.country, "NaN");
    strcpy(s.date_added, "March 1, 1900");
    s.release_year = 0;
    strcpy(s.rating, "NaN");
    strcpy(s.duration, "NaN");
    s.listed_in_size = 0;
    return s;
}

// Função para ler uma linha CSV e preencher a struct Show
Show ler_show(char* linha) {
    Show s = criar_show_vazio();
    char* token;
    char* ptr = linha;
    char campo[12][MAX_LINE] = {{0}};
    int i = 0;
    bool aspas = false;
    int idx = 0;
    int campo_idx = 0;

    while (*ptr) {
        if (*ptr == '"') {
            aspas = !aspas;
        } else if (*ptr == ',' && !aspas) {
            campo[campo_idx][idx] = '\0';
            campo_idx++;
            idx = 0;
        } else {
            campo[campo_idx][idx++] = *ptr;
        }
        ptr++;
    }
    campo[campo_idx][idx] = '\0';

    strcpy(s.show_id, strlen(campo[0]) ? campo[0] : "NaN");
    strcpy(s.type, strlen(campo[1]) ? campo[1] : "NaN");
    strcpy(s.title, strlen(campo[2]) ? campo[2] : "NaN");
    strcpy(s.director, strlen(campo[3]) ? campo[3] : "NaN");

    // Cast
    s.cast_size = 0;
    if (strlen(campo[4])) {
        token = strtok(campo[4], ",");
        while (token != NULL && s.cast_size < MAX_LIST) {
            while (*token == ' ') token++;
            strcpy(s.cast[s.cast_size++], token);
            token = strtok(NULL, ",");
        }
    }

    strcpy(s.country, strlen(campo[5]) ? campo[5] : "NaN");
    strcpy(s.date_added, strlen(campo[6]) ? campo[6] : "March 1, 1900");
    s.release_year = atoi(campo[7]);
    strcpy(s.rating, strlen(campo[8]) ? campo[8] : "NaN");
    strcpy(s.duration, strlen(campo[9]) ? campo[9] : "NaN");

    // Listed_in
    s.listed_in_size = 0;
    if (strlen(campo[10])) {
        token = strtok(campo[10], ",");
        while (token != NULL && s.listed_in_size < MAX_LIST) {
            while (*token == ' ') token++;
            strcpy(s.listed_in[s.listed_in_size++], token);
            token = strtok(NULL, ",");
        }
    }

    return s;
}

// Função para imprimir um Show
void imprimir_show(Show s) {
    printf("=> %s ## %s ## %s ## %s ## [", s.show_id, s.title, s.type, s.director);
    for (int i = 0; i < s.cast_size; i++) {
        printf("%s", s.cast[i]);
        if (i < s.cast_size - 1) printf(", ");
    }
    printf("] ## %s ## %s ## %d ## %s ## %s ## [", s.country, s.date_added, s.release_year, s.rating, s.duration);
    for (int i = 0; i < s.listed_in_size; i++) {
        printf("%s", s.listed_in[i]);
        if (i < s.listed_in_size - 1) printf(", ");
    }
    printf("] ##\n");
}

// Inserção na pilha
No* inserir_pilha(No* topo, Show s) {
    No* novo = (No*)malloc(sizeof(No));
    novo->elemento = s;
    novo->proximo = topo;
    return novo;
}

// Remoção da pilha
No* remover_pilha(No* topo, Show* removido) {
    if (topo == NULL) return NULL;
    *removido = topo->elemento;
    No* temp = topo;
    topo = topo->proximo;
    free(temp);
    return topo;
}

int tamanho_pilha(No* topo) {
    int tamanho = 0;
    while (topo != NULL) {
        tamanho++;
        topo = topo->proximo;
    }
    return tamanho;
}

void imprimir_pilha_com_indice(No* topo, int indice) {
    if (topo != NULL) {
        printf("[%d] ", indice);
        imprimir_show(topo->elemento);
        imprimir_pilha_com_indice(topo->proximo, indice - 1);
    }
}


// Busca por linha no arquivo
char* buscar_linha_por_id(const char* id, FILE* arquivo) {
    static char linha[MAX_LINE];
    rewind(arquivo);
    fgets(linha, MAX_LINE, arquivo); // pula cabeçalho

    while (fgets(linha, MAX_LINE, arquivo)) {
        char id_atual[MAX_FIELD];
        sscanf(linha, "%[^,]", id_atual);
        if (strcmp(id, id_atual) == 0) {
            return linha;
        }
    }
    return NULL;
}

int main() {
    char entrada[MAX_FIELD];
    char* linhas[1000];
    int total_ids = 0;
    No* pilha = NULL;

    while (scanf(" %[^\n]", entrada) && strcmp(entrada, "FIM") != 0) {
        linhas[total_ids] = strdup(entrada);
        total_ids++;
    }

    FILE* arquivo = fopen("/tmp/disneyplus.csv", "r");
    if (!arquivo) {
        printf("Arquivo não encontrado\n");
        return 1;
    }

    // Inserção inicial
    for (int i = 0; i < total_ids; i++) {
        char* linha = buscar_linha_por_id(linhas[i], arquivo);
        if (linha) {
            Show s = ler_show(linha);
            pilha = inserir_pilha(pilha, s);
        }
        free(linhas[i]);
    }

    int n;
    scanf("%d", &n);
    getchar(); // consumir newline

    for (int i = 0; i < n; i++) {
        fgets(entrada, sizeof(entrada), stdin);
        entrada[strcspn(entrada, "\n")] = '\0';

        if (entrada[0] == 'I') {
            char id[MAX_FIELD];
            sscanf(entrada, "I %s", id);
            char* linha = buscar_linha_por_id(id, arquivo);
            if (linha) {
                Show s = ler_show(linha);
                pilha = inserir_pilha(pilha, s);
            }
        } else if (entrada[0] == 'R') {
            Show removido;
            pilha = remover_pilha(pilha, &removido);
            printf("(R) %s\n", removido.title);
        }
    }
    int tamanho = tamanho_pilha(pilha);
    imprimir_pilha_com_indice(pilha, tamanho - 1);
    

    fclose(arquivo);
    return 0;
}

File: Aeds2/TP02/Disney/DisneyPilha.java
import java.util.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.Date;

class Show implements Cloneable {
    private String show_id;
    private String type;
    private String title;
    private String director;
    private List<String> cast = new ArrayList<>();
    private String country;
    private Date date_added;
    private int release_year;
    private String rating;
    private String duration;
    private List<String> listed_in = new ArrayList<>();

    private static final SimpleDateFormat sdf = new SimpleDateFormat("MMMM d, yyyy", Locale.ENGLISH);

    public Show() {
        this.show_id = "NaN";
        this.type = "NaN";
        this.title = "NaN";
        this.director = "NaN";
        this.cast = new ArrayList<>();
        this.country = "NaN";
        try {
            this.date_added = sdf.parse("March 1, 1900");
        } catch (ParseException e) {
            this.date_added = new Date();
        }
        this.release_year = 0;
        this.rating = "NaN";
        this.duration = "NaN";
        this.listed_in = new ArrayList<>();
    }

    public Show(String show_id, String type, String title, String director, String[] cast, String country,
                Date date_added, int release_year, String rating, String duration, String[] listed_in) {
        this.show_id = show_id;
        this.type = type;
        this.title = title;
        this.director = director;
        this.cast = new ArrayList<>(Arrays.asList(cast));
        this.country = country;
        this.date_added = date_added;
        this.release_year = release_year;
        this.rating = rating;
        this.duration = duration;
        this.listed_in = new ArrayList<>(Arrays.asList(listed_in));
    }

    public void ler(String linha) {
        String[] partes = new String[12];
        List<String> campos = new ArrayList<>();
        boolean aspas = false;
        StringBuilder campo = new StringBuilder();

        for (int i = 0; i < linha.length(); i++) {
            char c = linha.charAt(i);
            if (c == '"') {
                aspas = !aspas;
            } else if (c == ',' && !aspas) {
                campos.add(campo.toString().trim());
                campo.setLength(0);
            } else {
                campo.append(c);
            }
        }
        campos.add(campo.toString().trim());

        for (int i = 0; i < partes.length; i++) {
            if (i < campos.size()) {
                partes[i] = campos.get(i).isEmpty() ? "NaN" : campos.get(i);
            } else {
                partes[i] = "NaN";
            }
        }

        this.show_id = partes[0];
        this.type = partes[1];
        this.title = partes[2];
        this.director = partes[3];

        this.cast = new ArrayList<>();
        if (!partes[4].equals("NaN")) {
            this.cast = new ArrayList<>(Arrays.asList(partes[4].split(", ")));
            Collections.sort(this.cast);
        }

        this.country = partes[5];
        try {
            this.date_added = partes[6].equals("NaN") ? sdf.parse("March 1, 1900") : sdf.parse(partes[6]);
        } catch (ParseException e) {
            try {
                this.date_added = sdf.parse("March 1, 1900");
            } catch (ParseException ex) {
                this.date_added = new Date();
            }
        }

        try {
            this.release_year = Integer.parseInt(partes[7]);
        } catch (NumberFormatException e) {
            this.release_year = 0;
        }

        this.rating = partes[8];
        this.duration = partes[9];

        this.listed_in = new ArrayList<>();
        if (!partes[10].equals("NaN")) {
            this.listed_in = new ArrayList<>(Arrays.asList(partes[10].split(", ")));
            Collections.sort(this.listed_in);
        }
    }

    public void imprimir() {
        String castStr = (cast.isEmpty()) ? "[NaN]" : cast.toString();
        String listedInStr = (listed_in.isEmpty()) ? "[NaN]" : listed_in.toString();
    
        System.out.println("=> " + show_id + " ## " + title + " ## " + type + " ## " + director + " ## " + castStr + " ## " +
                country + " ## " + sdf.format(date_added) + " ## " + release_year + " ## " + rating + " ## " +
                duration + " ## " + listedInStr + " ##");
    }
    

    public Show clone() {
        try {
            Show clone = (Show) super.clone();
            clone.cast = new ArrayList<>(this.cast);
            clone.listed_in = new ArrayList<>(this.listed_in);
            clone.date_added = (Date) this.date_added.clone();
            return clone;
        } catch (CloneNotSupportedException e) {
            return null;
        }
    }

    public String getTitle() {
        return title;
    }
}

class No {
    public Show elemento;
    public No proximo;

    public No(Show elemento) {
        this.elemento = elemento;
        this.proximo = null;
    }
}

class PilhaShow {
    private No topo;

    public PilhaShow() {
        topo = null;
    }

    public void inserir(Show show) {
        No novo = new No(show);
        novo.proximo = topo;
        topo = novo;
    }

    public Show remover() {
        if (topo == null) {
            return null;
        }
        Show removido = topo.elemento;
        topo = topo.proximo;
        return removido;
    }

    public int tamanhoPilha() {
        int tamanho = 0;
        No aux = topo;
        while (aux != null) {
            tamanho++;
            aux = aux.proximo;
        }
        return tamanho;
    }

    public void imprimirPilhaComIndice() {
        int tamanho = tamanhoPilha();
        imprimirRecursivo(topo, tamanho - 1);
    }

    private void imprimirRecursivo(No no, int indice) {
        if (no != null) {
            System.out.print("[" + indice + "] ");
            no.elemento.imprimir();
            imprimirRecursivo(no.proximo, indice - 1);
        }
    }
}

public class DisneyPilha {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String> idsProcurados = new ArrayList<>();
        String entrada;

        while (!(entrada = scanner.nextLine()).equals("FIM")) {
            idsProcurados.add(entrada.trim());
        }

        PilhaShow pilha = new PilhaShow();

        Map<String, String> mapaLinhaPorId = new HashMap<>();
        try {
            Scanner arquivo = new Scanner(new File("/tmp/disneyplus.csv"));
            arquivo.nextLine(); 

            while (arquivo.hasNextLine()) {
                String linha = arquivo.nextLine();
                String id = linha.split(",")[0];
                mapaLinhaPorId.put(id, linha);
            }
            arquivo.close();
        } catch (FileNotFoundException e) {
            System.out.println("Arquivo não encontrado");
            
        }

        for (String id : idsProcurados) {
            if (mapaLinhaPorId.containsKey(id)) {
                Show s = new Show();
                s.ler(mapaLinhaPorId.get(id));
                pilha.inserir(s);
            }
        }

        int n = Integer.parseInt(scanner.nextLine());

        for (int i = 0; i < n; i++) {
            String comando = scanner.nextLine();
            String[] partes = comando.split(" ");

            switch (partes[0]) {
                case "I":  
                    if (mapaLinhaPorId.containsKey(partes[1])) {
                        Show s = new Show();
                        s.ler(mapaLinhaPorId.get(partes[1]));
                        pilha.inserir(s);
                    }
                    break;

                case "R":  
                    Show removido = pilha.remover();
                    if (removido != null) {
                        System.out.println("(R) " + removido.getTitle());
                    }
                    break;
            }
        }

        pilha.imprimirPilhaComIndice();

        scanner.close();
    }
}
